### ISSUE TITLE
API endpoints issues

### ISSUE THREAD

**Issue by reviloera** [tananaev/traccar#1437]
when I make a <code>GET</code>  request on <code>/api/device/get</code>  Am getting this error
<code>{"success":false,"error":"User not logged in - AccessControlException (BaseServlet:60 < DeviceServlet:49 < *:30 < BaseServlet:47 < ...)"}</code>

Yet I have loged in :

Again when I make a <code>GET</code>  request on <code>/api/async</code>  I am not getting any results back i.e  it doest finish executing.

Am using angularjs Restangular library 

Thank you .

**Comment by tananaev** [comment]
You need to login first before you can call other APIs:

https://www.traccar.org/traccar-api/

Async API is using long polling (https://en.wikipedia.org/wiki/Push_technology#Long_polling).

**Comment by reviloera** [comment]
I have already logged in and still getting that error message.

**Comment by tananaev** [comment]
Are you sure you are using same session?

**Comment by reviloera** [comment]
I have realized what is causig the problem 

Am not sending headers with credential and my cross origin at the server is *  , am trying to change then test if it works am updating you

**Comment by reviloera** [comment]
<code>
String clientOrigin = req.getHeader("origin");

            resp.setContentType("application/json");
            resp.setCharacterEncoding("UTF-8");
            resp.setHeader("Access-Control-Allow-Origin", clientOrigin);
            resp.setHeader("Access-Control-Allow-Credentials", "true");
	    resp.setHeader("Access-Control-Allow-Methods", "POST, GET, PUT, OPTIONS, DELETE");		
	    resp.setHeader("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
</code>

**Comment by reviloera** [comment]
Everything is okay now after the change above . Now logout url?

**Comment by tananaev** [comment]
Latest source code already has following code:
<pre>resp.setHeader("Access-Control-Allow-Origin", "*");
resp.setHeader("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
resp.setHeader("Access-Control-Allow-Methods", "GET, POST");</pre>

I believe it should be enough for CORS.

**Comment by reviloera** [comment]
If your client is on a different domain then Origin can't be *, you will get an error. And from the client I have to send my credentials, therefore, the server must allow for that. That is my I did the modifications otherwise it can't work. Am using Angularjs on the client.

**Comment by reviloera** [comment]
*That is why ...

**Comment by tananaev** [comment]
Quote from MDN:
> `Access-Control-Allow-Origin: *` which means that the resource can be accessed by any domain in a cross-site manner

What credentials are you sending to the server?

**Comment by vipien** [comment]
Hello, I have the same problem. I need to send commands with API and Browser response " CORS 'Access-Control-Allow-Origin' faltante ".

I have the lastest version Traccar and this is my code `function setCommand(){


	var att = JSON.stringify( {
		type:"positionPeriodic",
		deviceId:4,
		id:-4,
		attributes:{
			"frequency":2
		}

	} );


	$.ajax({
		url:url+"commands",
		type: "post",
		 contentType: "application/json",
		dataType: "json",
		processData : false,
		data: att,
		xhrFields: {
	       withCredentials: true
	    },
	    crossDomain: true,
		success:function(data) {
			console.log(data)
		}

	});`

Please help me!

**Comment by tananaev** [comment]
Check the response headers.

**Comment by vipien** [comment]
I have this line in traccar.xml `    <entry key='web.origin'>http://localhost:1080</entry>. But I can't acces.

**Comment by vipien** [comment]
I have this line in traccar.xml 

`    <entry key='web.origin'>http://localhost:1080</entry>
`

**Comment by tananaev** [comment]
Try to remove the line. By default all origins are allowed. Are you sure you include correct origin in the request headers?

**Comment by vipien** [comment]
Thank you very much for your reply.


@tananaev  I'm trying to run the ajax from html I have a xampp server running on port 1080 on the same PC. Error attached a picture where you can see the origin.

![errorcors](https://cloud.githubusercontent.com/assets/8376694/12919485/72eab35e-cf16-11e5-8a20-d304af2c2e5f.PNG)


Remove the line and does not work. I do not understand when you say " Are you sure you include correct origin in the request headers? " 

Not if it's possible that my problem is that I have compiled the latest version of the repository because the installer that is on page does not work with the driver msqql

I tried with all browsers. Ajax I'm sending this in a previous post in this thread

Really Thanks.

**Comment by tananaev** [comment]
Is it possible to get your code to try on my local machine or get access to your environment to test?

**Comment by vipien** [comment]
@tananaev  Of course. If you want you can access my pc with TeamViewer or Anydesk and try it directly from my computer. Only tell me what day and what time you can.

Thanks.

**Comment by tananaev** [comment]
@vipien, please send me details by email. Address is in profile.
